### PR TITLE
Initial implementation of UMAPINFO

### PR DESCRIPTION
**Pull request by rfomin**
Based on libretro [PrBoom fork](https://github.com/libretro/libretro-prboom) and PrBoom+2.6um. Demo extension not implemented.
Tested on:
1. Fork in the Road [[doomworld](https://www.doomworld.com/forum/topic/107191-fork-in-the-road-a-umapinfo-project-2019-07-03-update-alpha-3/)]
2. DBP25: Dead, But Dreaming [[idgames](https://www.doomworld.com/idgames/levels/doom2/Ports/d-f/dbp25_dbd)]
3. Sigil v1.2
4. Several UMAPINFO files I made: [umapinfo.zip](https://github.com/fabiangreffrath/woof/files/6172858/umapinfo.zip)

### CODE REVIEW

File: Source/u_scanner.h [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+boolean      U_HasTokensLeft(u_scanner_t* scanner);
+boolean      U_MustGetToken(u_scanner_t* scanner, char token);
+boolean      U_MustGetIdentifier(u_scanner_t* scanner, const char *ident);
+boolean      U_MustGetInteger(u_scanner_t* s);
+boolean      U_MustGetFloat(u_scanner_t* s);
+boolean      U_CheckToken(u_scanner_t* scanner, char token);
+boolean      U_CheckInteger(u_scanner_t* s);
+boolean      U_CheckFloat(u_scanner_t* s);
+
+void         U_Error(u_scanner_t* s, const char *msg, ...);
```
Comment (fabiangreffrath) [inline Source/u_scanner.h:104]: You will need to flag this function with a `PRINTF_ATTR` macro. Have a look at m_misc2.h.

File: Source/f_finale.c [hunk 2]
```diff
@@ -302,10 +372,19 @@ void F_TextWrite (void)
   int         cx;
   int         cy;
   
+  // [FG] if interbackdrop does not specify a valid flat, draw it as a patch instead
```
Comment (fabiangreffrath) [inline Source/f_finale.c:375]: This code looks familiar. So, we are supporting Rev 1.3 of the spec. Could you please add the UMAPINFO document from prboom-plus to the docs, with an install rule?

**Review by fabiangreffrath — APPROVED** [review]
Awesome, thank you! Though, this will unfortunately have to wait until the 4.0.2 release. I want to get rid of the UMAPINFO savegame bug before.

**Comment by rfomin** [comment]
> UMAPINFO savegame bug

Did you mean MUSINFO? I can't replicate that bug again.

File: Source/m_menu.c [hunk 3]
```diff
@@ -6027,6 +6027,7 @@ void M_Init(void)
       MainMenu[readthis] = MainMenu[quitdoom];
       MainDef.numitems--;
       MainDef.y += 8;
+      if (!EpiCustom)
```
Comment (fabiangreffrath) [inline Source/m_menu.c:6030]: Brackets, please.

File: Source/f_finale.c [hunk 4]
```diff
@@ -842,8 +842,12 @@ void F_Drawer (void)
     if (!finalestage || !gamemapinfo->endpic[0] || (strcmp(gamemapinfo->endpic, "-") == 0))
     {
       F_TextWrite();
-      return;
     }
+    else
```
Comment (fabiangreffrath) [inline Source/f_finale.c:846]: Does this deviate from the reference implementation?

Comment (rfomin) [reply Source/f_finale.c:846]: No, now it matches with the reference implementation:
https://github.com/coelckers/prboom-plus/blob/a43dff0255b6d81c5115010f27cd65b2598ded98/prboom2/src/f_finale2.c#L164-#L169

Comment (fabiangreffrath) [reply Source/f_finale.c:846]: See, thanks!

**Comment by fabiangreffrath** [comment]
So, by now we support Rev 1.4 of the spec. :wink: Could you update the docs again, please?

**Comment by fabiangreffrath** [comment]
Thank you very much!

**Pull request merged**